pkg/buses/message_dispatcher_test.go

```diff
@@ -0,0 +1,344 @@
+				Body:       ioutil.NopCloser(bytes.NewBufferString("destination-response")),
+			},
+			expectedErr: true,
+		},
+		"destination and reply - dest returns empty body": {
+			destination: "test-destination-svc.test-namespace.svc.cluster.local",
+			replyTo:     "test-reply-svc.test-namespace.svc.cluster.local",
+			message: &Message{
+				Headers: map[string]string{
+					// do-not-forward should not get forwarded.
+					"do-not-forward": "header",
+					"x-request-id":   "id123",
+					"knative-1":      "knative-1-value",
+					"knative-2":      "knative-2-value",
+					"ce-abc":         "ce-abc-value",
+				},
+				Payload: []byte("destination"),
+			},
+			expectedDestRequest: &requestValidation{
+				url: "http://test-destination-svc.test-namespace.svc.cluster.local/",
+				headers: map[string][]string{
+					"x-request-id": {"id123"},
+					"knative-1":    {"knative-1-value"},
+					"knative-2":    {"knative-2-value"},
+					"ce-abc":       {"ce-abc-value"},
+				},
+				body: "destination",
+			},
+			fakeResponse: &http.Response{
+				StatusCode: http.StatusAccepted,
+				Header: map[string][]string{
+					"do-not-passthrough": {"no"},
+					"x-request-id":       {"altered-id"},
+					"knative-1":          {"new-knative-1-value"},
+					"ce-abc":             {"new-ce-abc-value"},
+				},
+				Body: ioutil.NopCloser(bytes.NewBufferString("")),
+			},
+		},
+		"destination and reply": {
+			destination: "test-destination-svc.test-namespace.svc.cluster.local",
+			replyTo:     "test-reply-svc.test-namespace.svc.cluster.local",
+			message: &Message{
+				Headers: map[string]string{
+					// do-not-forward should not get forwarded.
+					"do-not-forward": "header",
+					"x-request-id":   "id123",
+					"knative-1":      "knative-1-value",
+					"knative-2":      "knative-2-value",
+					"ce-abc":         "ce-abc-value",
+				},
+				Payload: []byte("destination"),
+			},
+			expectedDestRequest: &requestValidation{
+				url: "http://test-destination-svc.test-namespace.svc.cluster.local/",
+				headers: map[string][]string{
+					"x-request-id": {"id123"},
+					"knative-1":    {"knative-1-value"},
+					"knative-2":    {"knative-2-value"},
+					"ce-abc":       {"ce-abc-value"},
+				},
+				body: "destination",
+			},
+			fakeResponse: &http.Response{
+				StatusCode: http.StatusAccepted,
+				Header: map[string][]string{
+					"do-not-passthrough": {"no"},
+					"x-request-id":       {"altered-id"},
+					"knative-1":          {"new-knative-1-value"},
+					"ce-abc":             {"new-ce-abc-value"},
+				},
+				Body: ioutil.NopCloser(bytes.NewBufferString("destination-response")),
+			},
+			expectedReplyRequest: &requestValidation{
+				url: "http://test-reply-svc.test-namespace.svc.cluster.local/",
+				headers: map[string][]string{
+					"x-request-id": {"altered-id"},
+					"knative-1":    {"new-knative-1-value"},
+					"ce-abc":       {"new-ce-abc-value"},
+				},
+				body: "destination-response",
+			},
+		},
```

Comment:
I think you need a fake response of nil to test this futher

Reply:
As the response to httpClient.Do()? Based on the same logic as above, I don't think the real client can return nil. I will add the test based on the response to the other comment.

Reply:
Nil should now be handled correctly, but I no longer have the httpDoer interface, so have no way to test it. Should I add the interface back so that I can have a response=nil test?

Reply:
ok thanks for checking!